README.md
# PayBreak/RPC

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

## Install

Via Composer

``` bash
$ composer require paybreak/rpc
```

## Usage

``` php
class MyApi
{
    use \PayBreak\Rpc\Api;

    protected function getActions()
    {
        return [
            'ping' => [self::class, 'ping'],
        ];
    }

    protected function authenticate()
    {
        return true;
    }

    protected function getRequestAction()
    {
        return \PayBreak\Rpc\Request::getParam('action');
    }

    protected function getRequestParams()
    {
        return (array) \PayBreak\Rpc\Request::getParam('params');
    }
    
    protected function ping(array $params)
    {
        return ['pong' => true, 'request' => $params];
    }
}

$obj = new MyApi();

$obj->executeCall();

```

Basic [demo](demo)

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email dev@paybreak.com instead of using the issue tracker.

## Credits

- [PayBreak][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/paybreak/rpc.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/PayBreak/rpc/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/paybreak/rpc.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/paybreak/rpc.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/paybreak/rpc.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/paybreak/rpc
[link-travis]: https://travis-ci.org/PayBreak/rpc
[link-scrutinizer]: https://scrutinizer-ci.com/g/PayBreak/rpc/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/PayBreak/rpc
[link-downloads]: https://packagist.org/packages/paybreak/rpc
[link-author]: https://github.com/PayBreak
[link-contributors]: ../../contributors


CHANGELOG.md
# 0.2.1
2016-01-24

## Features
- `Validation::processDateParam()` added `$default` param

# 0.1.1
2016-01-24

## Bug Fixes
- Updated *composer* dependencies

# 0.1.0
2016-01-24

## Initial Release


tests/ApiTest.php
<?php
/*
 * This file is part of the PayBreak/basket package.
 *
 * (c) PayBreak <dev@paybreak.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace PayBreak\Rpc\Test;

use AspectMock\Test as test;
use PayBreak\Rpc\Api;

/**
 * Request Test
 *
 * @author WN
 */
class TestTest extends \PHPUnit_Framework_TestCase
{
    use Api;

    private $auth = true;
    private $action = 'none';
    private $params = [];

    protected function tearDown()
    {
        test::clean(); // remove all registered test doubles
        parent::tearDown();
    }

    protected function getActions()
    {
        return [
            'test'  => [self::class, 'fakeAction'],
            'none'  => [self::class, 'nonExisting'],
            'exc'   => [self::class, 'actionException'],
            'str'   => [self::class, 'actionString'],
        ];
    }

    protected function authenticate()
    {
        return $this->auth;
    }

    protected function getRequestAction()
    {
        return $this->action;
    }

    protected function getRequestParams()
    {
        return $this->params;
    }

    private function fakeAction(array $params)
    {
        return $params;
    }

    private function actionException(array $params)
    {
        throw new \Exception('Test 123');
    }

    private function actionString()
    {
        return 'acdf';
    }

    public function testNotAuthenticated()
    {
        $this->auth = false;
        $this->setExpectedException('PayBreak\Rpc\ApiException', 'Authentication failed', 401);
        $this->executeAction([self::class, 'fakeAction']);
    }

    public function testWrongParamas()
    {
        $this->params = 'dsdsd';
        $this->setExpectedException('PayBreak\Rpc\ApiException', 'Params must be an array', 422);
        $this->executeAction([self::class, 'fakeAction']);
    }

    public function testNonExistingAction()
    {
        $this->setExpectedException('PayBreak\Rpc\ApiException', 'Non existing action', 400);
        $this->getAction('sss');
    }

    public function testMethodNotFound()
    {
        $this->setExpectedException('PayBreak\Rpc\ApiException', 'Method not found', 400);
        $this->getAction('none');
    }

    public function testGetAction()
    {
        $this->assertCount(2, $this->getAction('test'));
    }

    public function testExecuteCall()
    {
        $this->params = ['aaa'];
        $this->action = 'test';
        $response = test::double('PayBreak\Rpc\Response', ['sendJson' => true]);

        $this->executeCall();

        $response->verifyInvokedOnce('sendJson', [['aaa'], 200]);
    }

    public function testExecuteErrorCall()
    {
        $this->params = ['aaa'];
        $this->action = 'test';
        $this->auth = false;
        $response = test::double('PayBreak\Rpc\Response', ['sendJson' => true]);

        $this->executeCall();

        $response->verifyInvokedOnce('sendJson', [['error' => 'Authentication failed'], 401]);
    }

    public function testExecuteExceptionCall()
    {
        $this->params = ['aaa'];
        $this->action = 'exc';
        $this->auth = true;
        $response = test::double('PayBreak\Rpc\Response', ['sendJson' => true]);

        $this->executeCall();

        $response->verifyInvokedOnce('sendJson', [['error' => 'Test 123'], 500]);
    }

    public function testWrongResponse()
    {
        $this->params = ['aaa'];
        $this->action = 'str';
        $this->auth = true;
        $response = test::double('PayBreak\Rpc\Response', ['sendJson' => true]);

        $this->executeCall();

        $response->verifyInvokedOnce('sendJson', [['error' => 'Unprocesable response'], 500]);
    }
}


tests/bootstrap.php
<?php
/*
 * Test Bootstrap
 */

/** @see https://github.com/Codeception/AspectMock */
include __DIR__.'/../vendor/autoload.php'; // composer autoload

$kernel = \AspectMock\Kernel::getInstance();
$kernel->init([
    'debug' => true,
    'includePaths' => [
        __DIR__.'/../src',
    ]
]);


tests/RequestTest.php
<?php
/*
 * This file is part of the PayBreak/basket package.
 *
 * (c) PayBreak <dev@paybreak.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace PayBreak\Rpc\Test;

use AspectMock\Test as test;
use PayBreak\Rpc\Request;

/**
 * Request Test
 *
 * @author WN
 */
class RequestTest extends \PHPUnit_Framework_TestCase
{
    protected function tearDown()
    {
        test::clean(); // remove all registered test doubles
        parent::tearDown();
    }

    public function testGetParam()
    {
        $func = test::func('PayBreak\Rpc', 'file_get_contents', json_encode(['xxx' => 'yyy']));
        $this->assertSame('yyy', Request::getParam('xxx'));
        $func->verifyInvoked();
    }

    public function testJson()
    {
        $this->assertSame(['xxx' => 'yyy'], Request::json());
    }

    public function testDefaultGetParam()
    {
        $this->assertNull(Request::getParam('dsd'));
        $this->assertSame('wew', Request::getParam('dsd', 'wew'));
    }

    public function testGet()
    {
        $_GET['xs'] = 125;
        $this->assertSame(125, Request::getParam('xs'));
    }

    public function testPost()
    {
        $_POST['xy'] = 1256;
        $this->assertSame(1256, Request::getParam('xy'));
    }
}


demo/index.php
<?php
/*
 * This file is part of the PayBreak/basket package.
 *
 * (c) PayBreak <dev@paybreak.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

/*
 * php -S localhost:8080 index.php
 */

include_once '../vendor/autoload.php';

class MyApi
{
    use \PayBreak\Rpc\Api;

    protected function getActions()
    {
        return [
            'ping' => [self::class, 'ping'],
        ];
    }

    protected function authenticate()
    {
        return true;
    }

    protected function getRequestAction()
    {
        return \PayBreak\Rpc\Request::getParam('action');
    }

    protected function getRequestParams()
    {
        return (array) \PayBreak\Rpc\Request::getParam('params');
    }

    protected function ping(array $params)
    {
        return ['pong' => true, 'request' => $params];
    }
}


$obj = new MyApi();

$obj->executeCall();


src/Request.php
<?php
/*
 * This file is part of the PayBreak/basket package.
 *
 * (c) PayBreak <dev@paybreak.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace PayBreak\Rpc;

/**
 * ApiRequest
 *
 * @author WN
 * @package PayBreak\Rpc
 */
final class Request
{
    private static $instance;

    private $body = [];

    /**
     * Returns the *Singleton* instance of this class.
     *
     * @return $this
     */
    public static function getInstance()
    {
        if (null === self::$instance) {
            self::$instance = new self();
        }

        return self::$instance;
    }

    private function __construct() {
        $this->body = $this->getRequest();
    }

    private function __clone() { }

    /**
     * @author WN
     * @return array
     */
    public static function json()
    {
        return self::getInstance()->body;
    }

    /**
     * @author WN
     * @param string $name
     * @param mixed|null $default
     * @return mixed
     */
    public static function getParam($name, $default = null)
    {
        $instance = self::getInstance();

        if (array_key_exists($name, $instance->body)) {
            return $instance->body[$name];
        }

        if (array_key_exists($name, $_POST)) {
            return $_POST[$name];
        }

        if (array_key_exists($name, $_GET)) {
            return $_GET[$name];
        }

        return $default;
    }

    /**
     * @author WN
     * @return array
     * @throws ApiException
     */
    private function getRequest()
    {
        $request = file_get_contents('php://input');
        $body = json_decode($request, true);

        if (is_array($body)) {

            return $body;
        }

        return [];
    }
}


src/Validation.php
<?php
/*
 * This file is part of the PayBreak/basket package.
 *
 * (c) PayBreak <dev@paybreak.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace PayBreak\Rpc;

use Carbon\Carbon;

/**
 * Validation
 *
 * @author WN
 * @package PayBreak\Bacs\Helper
 */
class Validation
{
    /**
     * @author WN
     * @param string $param
     * @param array $params
     * @return bool
     */
    public static function checkParamExistsAndNotEmpty($param, array $params)
    {
        return (bool) (array_key_exists($param, $params) && $params[$param] !== '');
    }

    /**
     * @author WN
     * @param string $param
     * @param array $params
     * @param string|null $message
     * @return mixed
     * @throws ApiException
     */
    public static function paramExists($param, array $params, $message = null)
    {
        if (!array_key_exists($param, $params)) {

            throw new ApiException(($message === null)?'Param ' . $param . ' is missing':$message, 422);
        }

        return $params[$param];
    }

    /**
     * @author WN
     * @param string $param
     * @param array $params
     * @param string|null $message
     * @return mixed
     * @throws ApiException
     */
    public static function paramExistsAndNotEmpty($param, array $params, $message = null)
    {
        self::paramExists($param, $params);

        if (!self::checkParamExistsAndNotEmpty($param, $params)) {

            throw new ApiException(($message === null)?'Param ' . $param . ' is empty':$message, 422);
        }

        return $params[$param];
    }

    /**
     * @param string $param
     * @param array $params
     * @param string|null $message
     * @return array
     * @throws ApiException
     */
    public static function paramExistsAndIsArray($param, array $params, $message = null)
    {
        self::paramExistsAndNotEmpty($param, $params);

        if (!is_array($params[$param])) {

            throw new ApiException(($message === null)?'Param ' . $param . ' is not an array':$message, 422);
        }

        return $params[$param];
    }

    /**
     * @param string $param
     * @param array $params
     * @param string|null $message
     * @return array
     * @throws ApiException
     */
    public static function paramExistsAndNotEmptyArray($param, array $params, $message = null)
    {
        self::paramExistsAndIsArray($param, $params);

        if (count($params[$param]) == 0) {

            throw new ApiException(($message === null)?'Param ' . $param . ' is an empty array':$message, 422);
        }

        return $params[$param];
    }

    /**
     * @author WN
     * @param string $param
     * @param array $params
     * @return int|float
     * @throws ApiException
     */
    public static function paramIsNumeric($param, array $params)
    {

        self::paramExistsAndNotEmpty($param, $params);

        if (!is_numeric($params[$param])) {

            throw new ApiException('Param ' . $param . ' must be numeric', 422);
        }

        return $params[$param];
    }

    /**
     * @author WN
     * @param string $param
     * @param array $params
     * @param Carbon|null $default
     * @return Carbon
     * @throws ApiException
     */
    public static function processDateParam($param, array $params, Carbon $default = null)
    {
        try {
            self::paramExistsAndNotEmpty($param, $params);
        } catch (ApiException $e) {
           return self::checkDefaultDate($e, $default);
        }

        self::isString($param, $params);

        try{
            $date = Carbon::parse($params[$param]);
        } catch (\Exception $e) {
            throw new ApiException('Param ' . $param . ' is not parsable date', 422, $e);
        }

        return $date;
    }

    /**
     * @author WN
     * @param ApiException $e
     * @param Carbon|null $default
     * @return Carbon
     * @throws ApiException
     */
    private static function checkDefaultDate(ApiException $e, Carbon $default = null)
    {
        if ($default instanceof Carbon) {
            return $default;
        }

        throw $e;
    }

    /**
     * @author WN
     * @param $param
     * @param array $params
     * @throws ApiException
     */
    private static function isString($param, array $params)
    {
        if (!is_string($params[$param])) {

            throw new ApiException('Param ' . $param . ' is in a wrong format', 422);
        }
    }
}


src/Api.php
<?php
/*
 * This file is part of the PayBreak/basket package.
 *
 * (c) PayBreak <dev@paybreak.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace PayBreak\Rpc;

/**
 * Trait Api
 *
 * @author WN
 * @package PayBreak\Rpc
 */
trait Api
{
    /**
     * Configuration of available actions
     *
     * 'action_name' => callable
     *
     * @author WN
     * @return callable[]
     */
    abstract protected function getActions();

    /**
     * Indicates if request is authenticated
     *
     * @author WN
     * @return bool
     */
    abstract protected function authenticate();

    /**
     * Return name of requested action
     *
     * @author WN
     * @return string
     */
    abstract protected function getRequestAction();

    /**
     * Return params to be passed to action
     *
     * @author WN
     * @return array
     */
    abstract protected function getRequestParams();

    /**
     * Handles request and sends Response. Default implementation.
     *
     * @author WN
     */
    public function executeCall()
    {
        try {
            $result = $this->executeAction($this->getAction($this->getRequestAction()));

            if (!is_array($result)) {
                throw new ApiException('Unprocesable response', 500);
            }

            Response::sendJson($result);

        } catch (ApiException $e) {

            Response::sendJson(['error' => $e->getMessage()], $e->getCode());

        } catch (\Exception $e) {

            Response::sendJson(['error' => $e->getMessage()], 500);
        }
    }

    /**
     * Get Callable of Action
     *
     * @param string $action
     * @return callable
     * @throws ApiException
     */
    protected function getAction($action)
    {
        $actions = self::getActions();

        if (!array_key_exists($action, $actions)) {

            throw new ApiException('Non existing action', 400);
        }

        if (!is_callable($actions[$action])) {
            throw new ApiException('Method not found', 400);
        }

        return $actions[$action];
    }

    /**
     * Execute Action
     *
     * Callable action must return an array or throw an exception
     *
     * @author WN
     * @param callable $action
     * @return array
     * @throws ApiException
     */
    protected function executeAction(callable $action)
    {
        if (!$this->authenticate()) {
            throw new ApiException('Authentication failed', 401);
        }

        if (!is_array($this->getRequestParams())) {
            throw new ApiException('Params must be an array', 422);
        }

        return call_user_func($action, $this->getRequestParams());
    }
}
